549 filled activity icons

jtklein opened the pull request:
This closes #549

In this PR I have added a call to the iNat API "observations/updates" to the main container of the MyObservations view. The logic for this endpoint was partly already included previously but removed during refactors of the MyObservation screen. I have changed the API call so that we also get data on whether the update that the user has not seen yet is a comment or an identification, thereby enabling us to show only the respective icon filled.

There is an expectation in an unrelated unit test to not call this API when the user is signed out, so I also added here that the API is only called when the currentUser object is truthy.

In the unit test for the new hook I have added an expectation to see is realm.write has been called in the appropriate places but have not actually added an expectation as to whether the actual flag has been changed in the local observation. I am not sure which one is preferable (and also not sure how I would test for the latter).

I introduced calling the observation/updates API with the useAuthenticatedQuery hook only when the currentUser is defined, however, would such a check not make sense to be part of useAuthenticatedQuery directly instead? In other words, an authenticated query should only happen when there is an authenticated.

albullington commented on src/realmModels/Observation.js:
```diff
@@ -473,7 +473,8 @@ class Observation extends Realm.Object {
       user: "User?",
       updated_at: "date?",
-      viewed: "bool?"
+      viewed_comments: "bool?",
```
does it make more sense to store this on the `Observation` or on the linked `Comment` or `Identification` model?

jtklein replied:
I would keep it in the Observation for now, because the only use case so far is if any one of the comments is not viewed. At the moment it does not matter which one exactly.

jtklein commented on src/components/MyObservations/MyObservationsContainer.js:
```diff
@@ -5,6 +5,7 @@ import React, { useState } from "react";
 import useCurrentUser from "sharedHooks/useCurrentUser";
 import useInfiniteScroll from "sharedHooks/useInfiniteScroll";
 import useLocalObservations from "sharedHooks/useLocalObservations";
+import useObservationsUpdates from "sharedHooks/useObservationsUpdates";
```
Sounds like a good idea to me. But in a different issue/PR I would say.

albullington commented:
@jtklein Do you have steps to take to try testing that this is working as expected? I tried logging into a different test account, leaving comments and ids on an observation from my main account, and then reloading MyObservations. I can see the number of ids and comments increasing, but I don't ever see filled icons. Maybe I'm missing something here?

jtklein commented:
Hm, that sounds pretty much like I tested it as well. If you can see the numbers increase, I am assuming both accounts are simultaneously on staging or production. There is an issue with the sync button not always reloading directly. If you close the app and open it back up does that work? (Or maybe I should have added something to the logic of the button as well?)

albullington commented:
Nope, it didn't work for me when I closed and reopened the app. Or when I tried navigating back to MyObs from other screens.

jtklein commented:
I have updated this PR now and I think it is again ready for review.
The information whether there is a new comment or identification is available is fetched from the server using a react-query hook in the MyObservations and ObsDetails screens. So, those two screens should have direct access to the server-side information whenever they rerender. I also added a refetch of the data whenever the "Sync" button is pressed.

On every app restart or coming into the foreground the realm state resets to all comments and identifications being viewed, because the information whether to show filled icon or not is stored in realm. This is necessary because the user might look at a new comment on the website and then expect the app to not show that there is something new.

kueda commented on tests/integration/ObsDetails.test.js:
```diff
@@ -0,0 +1,70 @@
+    jest.clearAllMocks();
+  } );
+
+  describe( "when there is an observation with an update", () => {
```
This is more like "when there is an observation where we don't know if the user has viewed comments". If you want to test the case when there is definitely a new comment, you would `expect( observation.viewed_comments ).toBe( false )`, right?

kueda reviewed (approved):
I don't think I found anything that needs to block merge, mostly just added some suggestions.

Here are a few errors I got in a first pass in iOS (16.2, iPhone 14 Pro simulator). Not sure if they're directly related to the changes in this PR or not:

I get this on first load of My Observations:

```
 ERROR  Warning: Failed prop type: Invalid props.style key `borderTopColor` supplied to `Image`.
```

That one seems to happen consistently on first load.

I got this after getting an update about an unviewed comment and tapping on that observation:

```
 ERROR  Warning: Cannot update a component (`MyObservationsContainer`) while rendering a different component (`ObsDetails`). To locate the bad setState() call inside `ObsDetails`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
    in ObsDetails (created by SceneView)
```

When I got to ObsDetails, the new comment did not appear (probably out of scope for this PR, also seems inconsistent).

jtklein commented on src/components/App.js:
```diff
@@ -43,6 +45,44 @@ const App = ( { children }: Props ): Node => {
+    // unsubscribe on unmount
+    return () => subscription.remove();
+  }, [onAppStateChange, setObservationUpdatesViewedRealm] );
+
```
Done

jtklein commented on src/components/MyObservations/ObsStatus.js:
```diff
@@ -23,30 +26,29 @@ const ObsStatus = ( {
   const numIdents = observation.identifications?.length || 0;
   const numComments = observation.comments?.length || 0;
+  const filledIdents = observation.viewed_identifications === false;
+  const filledComments = observation.viewed_comments === false;
```
Changed

jtklein commented on src/components/ObsDetails/ObsDetails.js:
```diff
@@ -88,17 +89,23 @@ const ObsDetails = (): Node => {
       onSuccess: () => {
         markViewedLocally();
         queryClient.invalidateQueries( ["fetchRemoteObservation", uuid] );
+        queryClient.invalidateQueries( ["fetchObservationUpdates"] );
```
Added

jtklein commented on src/sharedHooks/useObservationsUpdates.js:
```diff
@@ -0,0 +1,109 @@
+    // If both comments and identifications are already unviewed, nothing to do here
+    if (
+      existingObs.viewed_comments === false
+      && existingObs.viewed_identifications === false
```
Done

jtklein commented on src/sharedHooks/useObservationsUpdates.js:
```diff
@@ -0,0 +1,109 @@
+    }
+    // If the update is a comment, set the observation's viewed_comments to false
+    if (
+      existingObs.viewed_comments === true
```
Changed
